### DISCUSSION TITLE
GNSS Network Processing and time of processing

### DISCUSSION THREAD

**Discussion by ParisaShafiei** [groops-devs/groops#128] [Q&A] [answered]
Dear Torsten, dear Sebastian, 

I have started to do the GNSS network processing for the year 2020. For this purpose, I downloaded the example scenarios, downloaded the input files like RINEX and BRDM files and run it on two different systems. 

Based on what I understood so far, I deem that there is an issue slowing down the speed of the last configuration file called "030groopsGnssProcessing.xml". According to the article titled "Processing of GNSS constellations and ground station networks using the raw observation approach", by Sebastian et al.,  it is mentioned that "**on a desktop computer with a quad-core CPU at 3.6 GHz and 16 of RAM, one iteration using the full network takes approximately 11 min with the settings described in this section**." However, the thing is I have done it on my computer with the following characteristics and core-network processing and it took more than two days. 

![image](https://user-images.githubusercontent.com/99748448/199463946-dfc1459d-3b53-4d05-947e-58d91d94618f.png)

Besides, we gave it a try on a another system with 25 cores and the speed was the same. You can see the start and end time:

**start: 2022-10-26 09:37:45**
**end: 2022-10-28 08:51:22**
**48 hours per day**

So, I am wondering if there is a bug or an issue in the example scenario that increased the time of processing or not. If it takes more than two days just for the core stations, I think it would be worse than that for the whole network of 1212 stations. 

Like always, thank you for your time and help. Your support is really appreciated. 

With Kind Regards, 
Parisa

**Comment by zhedumi** [comment]
Hi Parisa,

With how many processors did you start the script and did you make sure that you set the Normal equations to be not kept in the memory within [selectNormalsBlockStructure](https://groops-devs.github.io/groops/html/gnssProcessingStepType.html#selectNormalsBlockStructure ) ? Otherwise the normalequations will overflow your ram and be written to your disc which will reduce the computation speed drastically.

Best regards,
Patrick

**Comment by ParisaShafiei** [comment]
Hi Patrick, 

Thank for your reply. I tried it on two machines. First, I used 8 processes with **groopsMpi (GNOME, 8 processes)** on my system and then used 25 processes with **"groopsMpi (GNOME, 25 processes)"** on another powerful machine. To my surprise, it took the same time on both of them. It should be noted that I am using the example scenario and I did not change the settings. I checked configuration file and the "selectNormalsBlockStructure" was disabled. Now, I enabled this program to see if can speed up the processing on our systems or not. I am also note sure why the speed was the same with 8 and 25 cores. I have added the zip file containing all the config files.Could you please check if there is any bug or issue in the config file? 

With Kind Regards, 
Parisa 

[GNSSNetProcessing.zip](https://github.com/groops-devs/groops/files/9922060/GNSSNetProcessing.zip)

**Reply by ParisaShafiei** [reply]
Hi Patrick, 

I have an update:

Like what you said, I enabled the "selectNormalsBlockStructure" and applied the following changes in the group processing step:
![image](https://user-images.githubusercontent.com/99748448/199683406-7e5717a7-59cb-4399-811a-6f72c6f28346.png)
As you can see in the attached log file, I lunched the config file yesterday at 17:25 and it is still in the third iteration. Obviously, the speed of processing did not changed at all. Please also note that this is the first step of network processing just with the core stations. 

[groops.log](https://github.com/groops-devs/groops/files/9926947/groops.log)

 Thank you in advance and I am looking forward to hearing from you as soon as possible. 

With Kind Regards, 
Parisa

**Reply by Waley04** [reply]
Hi,

I use Dell (Precision 7920 Tower Workstation) with a very high computational capability. However, I experience the same issues — 

The solution to this will assist me as well.

Thanks.

**Reply by FFFFFG-IGG** [reply]
> Hi Patrick,
> 
> Thank for your reply. I tried it on two machines. First, I used 8 processes with **groopsMpi (GNOME, 8 processes)** on my system and then used 25 processes with **"groopsMpi (GNOME, 25 processes)"** on another powerful machine. To my surprise, it took the same time on both of them. It should be noted that I am using the example scenario and I did not change the settings. I checked configuration file and the "selectNormalsBlockStructure" was disabled. Now, I enabled this program to see if can speed up the processing on our systems or not. I am also note sure why the speed was the same with 8 and 25 cores. I have added the zip file containing all the config files.Could you please check if there is any bug or issue in the config file?
> 
> With Kind Regards, Parisa
> 
> [GNSSNetProcessing.zip](https://github.com/groops-devs/groops/files/9922060/GNSSNetProcessing.zip)

Hi Parisa
I'd like to ask you how to run groops with more than 4 processes? I'm windows user and can get the only choice "groopsMpi(windows, 4 processes)":
![image](https://user-images.githubusercontent.com/59271400/228105073-8533d4c2-aa99-4732-b9cc-41b9bce81490.png)
Thus, I wonder what have you done to change the available number of processes? By using command line or changing settings of groops elsewhere?
Looking forward to your reply!
Yongzhao Fan

**Reply by ParisaShafiei** [reply]
Hi Yongzhao, 

You can change the number of processes allocated for your processing. For example in my case, I used 28 cores out of 56 for mt processing. To do so, you need to go to Settings -> Commands -> Create a new Label and Command based on your desire. 
![image](https://user-images.githubusercontent.com/99748448/229487449-9d23c5a7-5830-44f7-b402-72c83af24fe4.png)

Just copy the **Label** and **Command** from previous lines and change the numbers. 

With kind regards, 
Parisa

**Comment by akvas** [comment, marked as answer]
Hi @ParisaShafiei and @Waley04,

my best guess is that the BLAS/LAPACK multithreading conflicts with running multiple MPI processes. You can try setting the thread count of OpenBlas to 1 by setting the environment variables
```
export OPENBLAS_NUM_THREADS=1
# and/or
export OMP_NUM_THREADS=1
```
as shown in the [respective documentation](https://github.com/xianyi/OpenBLAS/blob/develop/USAGE.md#how-can-i-use-openblas-in-multi-threaded-applications) and see whether that solves the performance issues.

Cheers,
Andreas

**Comment by ParisaShafiei** [comment]
Dear Andreas, 

Thank you so so much for your great help. I was struggling with the mentioned issue in the last few weeks. I did what you said and lunched the config file again. As you see in the attached log file, it only took 1 hour and 43 minutes to finish core network processing. 

Thanks again for your great help. It's really appreciated. 

[groops-test.log](https://github.com/groops-devs/groops/files/9931193/groops-test.log)

With Kind Regards, 
Parisa

**Reply by FFFFFG-IGG** [reply]
> Dear Andreas,
> 
> Thank you so so much for your great help. I was struggling with the mentioned issue in the last few weeks. I did what you said and lunched the config file again. As you see in the attached log file, it only took 1 hour and 43 minutes to finish core network processing.
> 
> Thanks again for your great help. It's really appreciated.
> 
> [groops-test.log](https://github.com/groops-devs/groops/files/9931193/groops-test.log)
> 
> With Kind Regards, Parisa

Excuse me, does core net processing mentioned above contain all product estimation(clock orbit static-position bias)? I also wonder how many stations do you use as core net?

**Reply by ParisaShafiei** [reply]
Hi @FFFFFG-IGG 

By network processing, you get the GNSS satellite products, i.e. precise orbit, clocks, signal biases and the static position of the stations included in your network. 

In my case, I used 69 stations out of ~ 650 as core stations. 

With kind regards, 
Parisa

**Comment by ParisaShafiei** [comment]
Dear developers, 

I have a question about the second step in GNSS network processing. 

For the first step of network processing, all the core network (64 stations) is processed following the cookbook and the example scenario. However, there is a problem in the second step. Actually according to the example scenario, for the second step, I excluded the station list containing the name of the core stations in processing and I changed the settings and parametrization to consider the transmitter parameters as known values, but I got an error message for no-net constraints in processing. 

**no stations contribute for no-net constraints must be selected for position estimation too  continue...**

I have tried several options, but I still have this problem. Even I deleted the no-net constraints for positions in the second step, but I could not succeed. I am really wondering why I got this error for the no-net constraints. According to the documents, we only need these constraints when we are estimating satellite parameters and station positions at the same time and not in the second step which is similar to PPP.

I have attached the gnss processing config file with all the changed settings. I would be pleased if you could take a look at it and help me to solve this error. 

[GnssProcessing.zip](https://github.com/groops-devs/groops/files/9950128/GnssProcessing.zip)

Thank you in advance and I am looking forward to hearing from you as soon as possible.

With Kind Regards, 
Parisa

**Comment by zhedumi** [comment]
Hi Parisa,

the provided .zip file is empty. If possible could you provide the .log file as well?
I am not entirely sure what you want to compute. 
Are you intending to process GnssNetwork: unknown sat+receiver, unknown sat+rec clocks, unknown sat+rec biases?
Or are you intending to process PPP: unkown receiver, unknown rec clocks, unknown rec biases?
If you just want to compute a daily static PPP position of a stationList you have to adapt it by making the staticPosition just unconstrained by setting it to 0. (I think this is the same issue https://github.com/groops-devs/groops/issues/63#issuecomment-1183076493) 
Also in that case I would suggest computing each station individually with [forEachReceiverSeparately](https://groops-devs.github.io/groops/html/gnssProcessingStepType.html#forEachReceiverSeparately). 


best regards,
patrick

**Reply by ParisaShafiei** [reply]
Hi Patrick, 

Sorry, I uploaded the zip file again. I also added the log file showing the error. 
Actually, my purpose is to run gnss network processing to get the satellite products (orbits, clocks and signal biases) for he year 2020 and compare the results with TUG and if it is successful I will processed to the year 2021. 
To answer your question, I should say, I am intending to process GnssNetwork: unknown sat+receiver, unknown sat+rec clocks, unknown sat+rec biases.
Therefore, I need to follow the three main steps in network processing. The first step is done and now I am at the second step to process non-core stations which I got that error and stock. 

With Kind Regards, 
Parisa 

[GnssProcessing.zip](https://github.com/groops-devs/groops/files/9950528/GnssProcessing.zip)

**Reply by zhedumi** [reply]
The error occurs due to the reason that within parametrization->staticPosition: selectNoNetReceivers you do not provide the individual station position files but one gridPosition.txt. Also the first processing step should be selecting only the core network for estimation.
Furthermore the script at the moment is a bit of a mix of PPP/Networkprocessing. Since you actually want to compute all the aspects typically for a network processing I would refer to the [cookbook gnssNetwork](https://groops-devs.github.io/groops/html/cookbook.gnssNetwork.html) especially the provided [network example](https://ftp.tugraz.at/outgoing/ITSG/groops/scenario/scenarioGnssNetwork.zip) and compare your parametrization and processingsteps with this. 
This provided script basically computes the network with all the unknowns as you mentioned.

**Reply by ParisaShafiei** [reply]
Thank you for your reply.

So, to select **selectNoNetReceivers**, I just can select each processed station positions from core network processing which is done in the previous step. Am I right? 
And for core stations, I have two doubts about the list of core stations.
 Is the following is showing the 64 core stations? 

[stationList.txt](https://github.com/groops-devs/groops/files/9952287/stationList.txt)

Or the core stations are the first station in each 69 group based on the stationList.repro3.core.tiered.txt?

**Reply by zhedumi** [reply]
Hi yes you could use them as the same basically within selectNoNetReceivers. Usually it is good to use stable stations such as the core IGS14 stations or something similar. 
Yes basically the stationList.txt you have is just the first column of stationList.repro3.core.tiered.txt. The other columns are just stations which can be used in case the first column is not available. Since your network is rather small the core network, nonet network and the evaluated stations are all the same.

In other scenarios these can differ. 
For example you can have a set of 800 stations to be evaluated in stationList.txt. This file you would set within receiver->stationList.
Then from these 800 you may select ~200 station for a nonet stationList and putt them in stationList.nonet.txt which you set it selectNonetReceivers. 
And from these 800 you may also create a stationList.core.txt containing your core network which you process first for example 70 stations to begin with. With these 3 stationLists you would adapt the provided network processing script with the Advanced steps within [cookbook](https://groops-devs.github.io/groops/html/cookbook.gnssNetwork.html) at the bottom to be processed.

**Reply by ParisaShafiei** [reply]
Thank for your clarification about the core stations in network processing. 

For the second step, like what you said, it equals to PPP and I should process each non core station separately, if it is true, how I can use "forEachReceiverSeparately" in my processing, because in "selectrecerievs" section, I only see these options: all, wildcard, file and exclude and I could not find this option. 

On the other hand for the no-net constraints, while I am processing non-core stations, I should use the processed core stations in the selectNoNetReceivers, is it right? As you see in the attached picture, I pointed to the list of core stations in **selectNoNetreceivers**and for the inputfileNoNetPositions, it is pointed to the static positions of these core stations processed in the previous step. 

![image](https://user-images.githubusercontent.com/99748448/200360981-2308c655-9bf3-4a20-8721-f9460567daa9.png)

**Reply by zhedumi** [reply]
Wtihin forEachReceiverSeparately you only need to exclude the core network stations as described in [cookbook](https://groops-devs.github.io/groops/html/cookbook.gnssNetwork.html) within the advanced section. So within exclude select the station list consisting of the core network since they were already computed. 
No that is not necessarely needed. For example in my recent computations I used the same core network as on the ftp while my nonet stations are around 200 selected stations. 
In my example I had a core network of 69 stations, a nonetstation list of 200 stations and evaluation stations of X stations. The nonetposition station list just adds the constraints to the normal equation using the given 200 stations from the nonetstation list. They basically define the reference frame we want to align our network with therefore I would suggest using IGS14 position files for the stations. So in my example this 8 steps were processed
1. Estimating  a float solution for the 69 stations core network station list
2. Resolve the ambiguities of the 69 stations core network station list.
3. Estimate a solution with the solved ambiguities for the 69 stations core network list.
4. Next process forEachReceiverSeparately excluding the core stations used in 1,2,3 therefore we evaluate for X - 69 stations
5. Compute a individual station float solution for X - 69 stations
6. Resolve the individual station ambiguities for X - 69 stations
7. Compute a solution with the fixed ambiguities for X -69 stations
8. Use all stations and all parameters in one big least squares adjustment to estimate the solution for X stations

So simply said step 1 to 3 solves the core network thus "fixing" our satellite orbits for ambiguity solution.
With step 4,5,6,7 we hang the other stations into the network by basically using PPP.
Step 1 - 7 are basically just there to solve the ambiguities and give better a priori values for the last estimation process. Within 8 we do a least squares with the solved ambiguities and all parameters.

**Reply by ParisaShafiei** [reply]
Dear Patrick, 

Since you replied to my question, I have worked on the three main steps to complete the network processing. So sorry, for my long silence in replying back your message. 

First, I completed the steps 1 to 3 and I got the clocks and orbits for the satellites and static positions and troposphere parameters for the core stations as well as the signal biases for both satellites and stations. 

After finishing step 1, I continued to step 2 , to do PPP for each non-core stations based on your explanation. So, I have used the following parametrization and settings for the PPP part:

**parametrization:**
ionoshphereSTEC 
ionosphereVTEC
staticPositions: with selecting the core stations available in the {groopsDataDir}/gnss/receiverStation/position/igs/igb14/staticPosition.{station}.dat by using the satationList.nonet.txt in the selectNoNetReceivers section. 
signalBiases: with using the estimated transmitter signal biases in the core network as input files to get the signalBiases for the receiver 
codeBiases
ambiguities: to just only estimateReceiverPhaseBias 
troposphere
clocks 
tecBiases
troposphere constraints 

**processingsteps:**
foreachreceiverseparetly: with excluding the list of core stations 
estimte
resolveambiguities 
estimate 
writeResults 

So, at the end of step 2, I got the clocks, troposphere, vtec, static positions and signal biases for all the non core stations. I hope the applied settings were correct. 

For the third step, it is still a question for me how I can get the final results without resolving the ambiguities. Indeed, by using the following settings, I got the orbits for the day 2020-08-01 and compared them with the TUG orbits. They differences are considerable and it seems there is a systematic error in the processing. I wanted to ask you if there is something wrong in the settings that I have applied to get the final outputs or not. For this purpose, I used the same group parametrization in core network processing and only applied the processing steps mentioned in the cookbook. 


![orbitDiff_2020-08-01 G01](https://user-images.githubusercontent.com/99748448/204014060-682e04c3-0bb7-43e2-888c-9de4852cceca.png)

![orbitDiff_2020-08-01 G02](https://user-images.githubusercontent.com/99748448/204014427-fbc0ce41-90d4-4dae-ae75-188f1cd35399.png)


Thanks a lot for your time. You have really helped me so far. 

Looking forward to hearing from you.

With kind regards, 
Parisa

**Reply by zhedumi** [reply]
Dear Parisa,

Maybe just one question from me since I am a bit confused. The steps I described are all run within the same groopsscript and the parametrization does not need to be adapted throughout it whether it computes the core network or the PPP solutions etc. The parametrization should be exactly as in the cookbook for network processing and your network processingsteps in the groops script then should look as in the following screenshot. 
![networkprocessing](https://user-images.githubusercontent.com/77068736/204349497-dfa8bc50-e4cd-4f65-b638-efb0fe244839.png)

One thing to note is for repro3 we selected a stricter no-net translation constraint so if you truly want to replicate the results you might want to change it form 0.001 to 0.00001 but there will still linger differences between the TUG final results and any new computed one. To truly compare the orbits you'd have to compute a helmert transformation to accommodate any geometrical changes due to different used GNSS, stations, observations etc. 
The difference in clock is also not really surprising and also a result of different used stations in my opinion. Also note the differences are basically smaller than 1 ns (1 ns ~ 30cm iirc).

Best regards,
Patrick

**Reply by ParisaShafiei** [reply]
Dear Patrick, 

First of all, I would like to express that I really appreciate your help. I did what you said and I got the results for the year 2020. I also compared them with the CODE orbits. I have also continued to process the year 2021 and use these new products in our processing. I had attached the three plots for the orbit differences in along, cross and radial. Regarding what you told me for the helmert transformation, I applied it for each orbit comparison that I did. 

PS: Even when we want to compare two orbits, e.g. TUG and AIUB of a LEO satellite, should we apply the helmert transformation as well before going to rotate the orbits from CRF to LOF? It is a inevitable step in any orbit comparisons in local orbit frame, am I right? 

![along_orbitDiff_2020-08-01 G01](https://user-images.githubusercontent.com/99748448/205888222-90934fda-c643-4c0f-a330-7a431a355b46.png)

![cross_orbitDiff_2020-08-01 G01](https://user-images.githubusercontent.com/99748448/205888264-c1c93121-6b57-4115-b742-0073f247b2be.png)

![radial_orbitDiff_2020-08-01 G01](https://user-images.githubusercontent.com/99748448/205888292-cff637be-d3c5-43d9-9e58-54f473ad1269.png)

Thanks again for your time and help. 

With kind regards, 
Parisa

**Reply by zhedumi** [reply]
Hi Parisa,

In a first check I would always suggest doing a helmert transformation between the orbits to check if there is a global systematic component (translations, rotation or scale). It could be that some 
orbits use different models or constraints leading tot hose differences and the helmert transformation parameters are a good indication on the differences. 
For example for a short research I compared the IGS rapid to the IGS final orbit products which would show such differences in the coordinates:
(x - axis are the days of the 8th month of the year 2022. This was just a fast plot.)
![orbitdiff original G01](https://user-images.githubusercontent.com/77068736/208376451-abec03bd-1f3c-4d63-aa0f-5409c341a355.png)
![orbitdiff original G19](https://user-images.githubusercontent.com/77068736/208377657-4705d1fc-7a07-4dfe-adac-47b5952412d6.png)
![orbitdiff original G28](https://user-images.githubusercontent.com/77068736/208377629-feff58f2-76bd-4cb5-a0e7-71015956225c.png)
Looking at the products tho https://igs.org/products/ the accuracy should be ~2.5 cm and therefore really close aligned to the final and not show these high differences in my oppinion.
Checking helmert parameters it becomes quiet obvious as to why this difference occures:

![igsrapid_igsfinal_est_helmert_rot](https://user-images.githubusercontent.com/77068736/208376877-c302437b-fbe8-46a5-b726-6bfaa0a0b111.png)
![igsrapid_igsfinal_est_helmert_scale](https://user-images.githubusercontent.com/77068736/208376881-add9a252-8ec6-4b2a-92b9-a4650fdab121.png)
![igsrapid_igsfinal_est_helmert_trans](https://user-images.githubusercontent.com/77068736/208376883-d8ac7a30-c125-4802-a8c4-3358504f8186.png)

So the difference between the rapid and final products of IGS seem to have a systematic difference in the rotation. (I am not a 100% sure as to why this occurs but it is there apparently).
Transforming the rapid to the final products with helmert leads to a difference which I would have regarded as closely aligned.

![orbitdiff transd G01](https://user-images.githubusercontent.com/77068736/208377157-051581d8-fe1b-469c-88ea-83a15b1c7e9f.png)

So yeah to compare the orbits I would always do a helmert transformation and check the helmert parameters.

Best regards,

Patrick

**Reply by ParisaShafiei** [reply]
Hi Patrick, 

Thank you for the clarification on the helmert transformation. By looking at the example plots, I have noticed the importance of doing the helmert transformation and checking the estimated parameters. 

With kind regards, 
Parisa

**Comment by ParisaShafiei** [comment]
Hi Patrick, 

I hope you are doing well. 

For the network processing, as we have already discussed about the essential steps and setting, I have notice two major changes that I would like to mention them here. 

First of all, I compared the signal biases with what we have in the original TUG products. For sure, I did not expect to get the same results, but what is really surprising for me is the that there are significant differences between both estimate code and phase biases, i.e. over 70 cm to 1m and 50 cm. On the other hand, I could not get the same phase signal types based on what has been written in the TUG signal biases. It seems those signal types are ignored in the processing steps. 

Secondly, when I was checking the station positions, I noticed that only core stations are processed and none of the non-core stations are processed. I have used the setting you have told me to get these final outputs and I wanted to check with you if you could get the whole the position of stations (X-60) at the end of processing. Please not that I have the RINEX observation data for about 600 station + 60 core station in total. 

Thank you for your time and help and I am really looking forward to hearing from you as soon as you have free time. 

With kind regards, 
Parisa

**Reply by zhedumi** [reply]
Hi Parisa,

The version we used to compute the repro3 final products was significantly different to the nowadays version. It is really complicated to actually compare the code and phase biases. This is due to the ionosphere, clock and biases being relative to each other. In TUG the written out biases were the **X and **Q biases and nowadays we write out the **I and **Q biases. 
The C?X biases can be created by the simple combinations since this is the base assumption we take:
C?X = C?C * 0.5 + 0.5 * C?B . the differences are normal and a result on our approach of the signal biases. In our approach we simulate the observation equations, creating the normal equations and proceed to create estimable linear combinations of the 
code biases by eigenvalue decomposition of the simulated normal equations. The eigenvectors are classified into a nullspace matrix and our transformation matrix responsible for generating the estimable linear combinations. For GPS we selectively set C1W and C2W to be our nullspace since this is IGS convention. For Galileo there is no such convention. And in Galileos case the stations around the globe either measure C?Q/C, or C?X which results in  "split network" leading to a group of receiver available for bias estimation for C?X and a group of receivers available for bias estimation of C?Q/C. This in combination of our pseudo absolute base estimation by null constraining can lead to abritary differences between I/Q. 
(More infos regarding the creation of the biases can be found here [Dumitraschkewitz 2022](https://files.igs.org/pub/resource/pubs/workshop/2022/IGSWS2022_S11_06_Dumitraschkewitz.pdf?_gl=1*1hfr0hv*_ga*MTE0NTQ2ODYyMi4xNjYzMjIzOTAz*_ga_Z5RH7R682C*MTY3MTQzNzAxOC4yNC4xLjE2NzE0MzgxNzMuNjAuMC4w&_ga=2.126309645.1834965583.1671437020-1145468622.1663223903)
[Strasser 2022](https://pure.tugraz.at/ws/portalfiles/portal/40917089/Strasser_2022_Doctoral_Thesis_Published_Version.pdf))

Comparing the signal biases is a very complicated matter entirely and they cannot be compared straight on see chapter 8.2.4 in [Strasser 2022](https://pure.tugraz.at/ws/portalfiles/portal/40917089/Strasser_2022_Doctoral_Thesis_Published_Version.pdf). So yeah the differences are not surprising.

Yes for me all stations are written out and I have all position files. 


Best regards,
Patrick

**Reply by ParisaShafiei** [reply]
Hi Patrick, 

Thanks for explaining clearly why the estimated signal biases are different and what is the reason behind these deviations. 

About the problem I had in processing the whole network and writing down all the position files, I found my mistake. In "receiver - stationNetwork" and then in "inputFileStationList", I just used the "stationList.repro3.core.tiered.txt" as the input and that was the reason I could not get all the station files of the whole network. So, by changing it to read the list of all stations, it worked properly.

With kind regards, 
Parisa

**Comment by FFFFFG-IGG** [comment]
Hello! I wonder if groopsMPI run different under windows and linux? I mean groopsMPI in windows can only run with 4 processes, does ubuntu can only run MPI with 4 processes too?

**Reply by ParisaShafiei** [reply]
Hi @FFFFFG-IGG ,

I am running my programs on Ubuntu and I use 28 processes in groopsMPI command. I think you can also change  the commands in Windows.

![image](https://user-images.githubusercontent.com/99748448/229489013-49624901-e54b-4112-810d-20bcd7f58593.png)


Hope it will help. 

With kind regards, 
Parisa

**Reply by FFFFFG-IGG** [reply]
> Hi @FFFFFG-IGG ,
> 
> I am running my programs on Ubuntu and I use 28 processes in groopsMPI command. I think you can also change the commands in Windows.
> 
> ![image](https://user-images.githubusercontent.com/99748448/229489013-49624901-e54b-4112-810d-20bcd7f58593.png)
> 
> Hope it will help.
> 
> With kind regards, Parisa

Hi Parisa,
First of all I'd like to aprreciate your help! Besides, I'm interest in processing time of per iteration when you processed the whole net(30s sampling) after core-net and ppp-AR, since it takes almost 2 hours with 4 prcessors, too much time, I think.
##settings in my config.xml:
parameters are set same as demo of TUG
system: GPS GALILEO BDS GLONASS
kind regards.
Yongzhao Fan

**Reply by ParisaShafiei** [reply]
Hi Yongzhao, 

For me, processing the whole network (i.e. ~ 650 stations) with only GPS, Galileo constellations taken into account, takes about 5 hours with writing the normal equations. 

This time that I mentioned is starting from the first step till the end of processing, i.e. (core, non core and whole network). 

With kind regards, 
Parisa
